### PR TITLE
Add enumerable category methods to NSArray

### PR DESCRIPTION
**Pull request by gfontenot**

### CODE REVIEW

File: BotKit/Categories/NSArray+Enumeration.m [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+{
+    NSMutableArray *returnArray = [NSMutableArray array];
+
+    for (id object in self) {
+        id returnObject = enumerationBlock(object);
+
+        if (returnObject) [returnArray addObject:returnObject];
+    }
+
+    return [NSArray arrayWithArray:returnArray];
```
Comment (hyperspacemark) [inline BotKit/Categories/NSArray+Enumeration.m:25]: Minor nit, this could just be `return [returnArray copy];`.

**Comment by hyperspacemark** [comment]
:thumbsup:

File: BotKit/Categories/NSArray+Enumeration.m (hunk 1 above)
Comment (gfontenot) [inline BotKit/Categories/NSArray+Enumeration.m:25]: Will `copy` return an immutable version? That was my intent here.

Comment (hyperspacemark) [inline BotKit/Categories/NSArray+Enumeration.m:25]: It will. Sending `-copy` to an instance of `NSMutableArray` will return an instance of `NSArray`.

Comment (gfontenot) [inline BotKit/Categories/NSArray+Enumeration.m:25]: Oh, cool. TIL. So then maybe this comes down to explicitness? I feel like using `+[NSArray arrayWithArray:]` is a bit more obvious about what it's returning.

Comment (hyperspacemark) [inline BotKit/Categories/NSArray+Enumeration.m:25]: Ya that's totally understandable. They're basically exactly the same under ARC anyway. Pre-ARC, the `-copy` approach would return a +1'd object while `+arrayWithArray:` would returned an autoreleased object.

**Comment by gfontenot** [comment]
Squashed and merged in efab620

**Pull request closed without merging**